File: lib/app/models/post_model.dart
// ignore_for_file: public_member_api_docs, sort_constructors_first
import 'dart:convert';

import 'package:flutter/foundation.dart';

import 'package:cartisan/app/models/delivery_options.dart';

class PostModel {
  String postId;
  String ownerId;
  String description;
  String productName;
  String brand;
  List<String> variants;
  double price;
  String location;
  double rating;
  List<String> images;
  String selectedVariant;
  int quantity;
  bool isProductInStock;
  bool archived;
  int sellCount;
  int commentCount;
  int reviewCount;
  int likesCount;
  DeliveryOptions deliveryOptions;

  int get priceInCents => (price * 100).toInt();

  PostModel({
    required this.postId,
    required this.ownerId,
    required this.description,
    required this.productName,
    required this.brand,
    required this.variants,
    required this.price,
    required this.location,
    required this.rating,
    required this.images,
    required this.selectedVariant,
    required this.quantity,
    required this.isProductInStock,
    required this.archived,
    required this.sellCount,
    required this.commentCount,
    required this.reviewCount,
    required this.likesCount,
    required this.deliveryOptions,
  });

  factory PostModel.fromMap(Map<String, dynamic> map) {
    return PostModel(
      postId: map['postId'] as String,
      ownerId: map['ownerId'] as String,
      description: map['description'] as String,
      productName: map['productName'] as String,
      brand: map['brand'] as String,
      variants: (map['variants'] as List<dynamic>).cast<String>(),
      price: (map['price'] as num).toDouble(),
      location: map['location'] as String,
      rating: (map['rating'] as num).toDouble(),
      images: (map['images'] as List<dynamic>).cast<String>(),
      selectedVariant: map['selectedVariant'] as String,
      quantity: map['quantity'] as int,
      isProductInStock: map['isProductInStock'] as bool,
      archived: map['archived'] as bool,
      sellCount: map['sellCount'] as int,
      commentCount: map['commentCount'] as int,
      reviewCount: map['reviewCount'] as int,
      likesCount: map['likesCount'] as int,
      deliveryOptions: DeliveryOptions.values[map['deliveryOptions'] as int],
    );
  }

  PostModel copyWith({
    String? postId,
    String? ownerId,
    String? description,
    String? productName,
    String? brand,
    List<String>? variants,
    double? price,
    String? location,
    double? rating,
    List<String>? images,
    String? selectedVariant,
    int? quantity,
    bool? isProductInStock,
    bool? archived,
    int? sellCount,
    int? commentCount,
    int? reviewCount,
    int? likesCount,
    DeliveryOptions? deliveryOptions,
  }) {
    return PostModel(
      postId: postId ?? this.postId,
      ownerId: ownerId ?? this.ownerId,
      description: description ?? this.description,
      productName: productName ?? this.productName,
      brand: brand ?? this.brand,
      variants: variants ?? this.variants,
      price: price ?? this.price,
      location: location ?? this.location,
      rating: rating ?? this.rating,
      images: images ?? this.images,
      selectedVariant: selectedVariant ?? this.selectedVariant,
      quantity: quantity ?? this.quantity,
      isProductInStock: isProductInStock ?? this.isProductInStock,
      archived: archived ?? this.archived,
      sellCount: sellCount ?? this.sellCount,
      commentCount: commentCount ?? this.commentCount,
      reviewCount: reviewCount ?? this.reviewCount,
      likesCount: likesCount ?? this.likesCount,
      deliveryOptions: deliveryOptions ?? this.deliveryOptions,
    );
  }

  Map<String, dynamic> toMap() {
    return <String, dynamic>{
      'postId': postId,
      'ownerId': ownerId,
      'description': description,
      'productName': productName,
      'brand': brand,
      'variants': variants,
      'price': price,
      'location': location,
      'rating': rating,
      'images': images,
      'selectedVariant': selectedVariant,
      'quantity': quantity,
      'isProductInStock': isProductInStock,
      'archived': archived,
      'sellCount': sellCount,
      'commentCount': commentCount,
      'reviewCount': reviewCount,
      'likesCount': likesCount,
      'deliveryOptions': deliveryOptions,
    };
  }

  String toJson() => json.encode(toMap());

  factory PostModel.fromJson(String source) =>
      PostModel.fromMap(json.decode(source) as Map<String, dynamic>);

  @override
  String toString() {
    return 'PostModel(postId: $postId, ownerId: $ownerId, description: $description, productName: $productName, brand: $brand, variants: $variants, price: $price, location: $location, rating: $rating, images: $images, selectedVariant: $selectedVariant, quantity: $quantity, isProductInStock: $isProductInStock, archived: $archived, sellCount: $sellCount, commentCount: $commentCount, reviewCount: $reviewCount, likesCount: $likesCount, deliveryOptions: $deliveryOptions)';
  }

  @override
  bool operator ==(covariant PostModel other) {
    if (identical(this, other)) return true;

    return other.postId == postId &&
        other.ownerId == ownerId &&
        other.description == description &&
        other.productName == productName &&
        other.brand == brand &&
        listEquals(other.variants, variants) &&
        other.price == price &&
        other.location == location &&
        other.rating == rating &&
        listEquals(other.images, images) &&
        other.selectedVariant == selectedVariant &&
        other.quantity == quantity &&
        other.isProductInStock == isProductInStock &&
        other.archived == archived &&
        other.sellCount == sellCount &&
        other.commentCount == commentCount &&
        other.reviewCount == reviewCount &&
        other.likesCount == likesCount &&
        other.deliveryOptions == deliveryOptions;
  }

  @override
  int get hashCode {
    return postId.hashCode ^
        ownerId.hashCode ^
        description.hashCode ^
        productName.hashCode ^
        brand.hashCode ^
        variants.hashCode ^
        price.hashCode ^
        location.hashCode ^
        rating.hashCode ^
        images.hashCode ^
        selectedVariant.hashCode ^
        quantity.hashCode ^
        isProductInStock.hashCode ^
        archived.hashCode ^
        sellCount.hashCode ^
        commentCount.hashCode ^
        reviewCount.hashCode ^
        likesCount.hashCode ^
        deliveryOptions.hashCode;
  }
}

File: functions/src/models/post_model.ts
import { DeliveryOptions, } from "./enums";

export class PostModel {
    postId:string;
    ownerId:string;
    username:string;
    description:string;
    productName:string;
    brand:string;
    variants:string[];
    price: number;
    location:string;
    rating:number;
    images:string[];
    reviewCount: number;
    likesCount: number;
    selectedVariant:string;
    deliveryOptions:DeliveryOptions;
    quantity:number;
    isProductInStock:boolean;
    archived: boolean;
    sellCount: number;
    commentCount: number;
    timestamp: number;
    constructor(
      {postId,ownerId,username,description,productName,brand,variants,price,location,rating,reviewCount,likesCount ,images,selectedVariant,
      deliveryOptions,quantity,isProductInStock,archived,sellCount, commentCount, timestamp}:
      {postId: string,
      ownerId:string,
      username:string,
      description:string,
      productName:string,
      brand:string,
      variants:string[],
      price: number,
      location:string,
      rating:number,
      images:string[],
      likesCount: number,
      reviewCount: number,
      selectedVariant:string,
      deliveryOptions:DeliveryOptions,
      quantity:number,
      isProductInStock:boolean,
      archived: boolean,
      sellCount: number, 
      commentCount: number
      timestamp: number
    }
    ){
      this.postId = postId;
      this.ownerId = ownerId;
      this.username = username;
      this.description = description;
      this.productName = productName;
      this.brand = brand;
      this.variants = variants;
      this.price = price;
      this.location = location;
      this.rating = rating;
      this.images = images;
      this.selectedVariant= selectedVariant;
      this.deliveryOptions = deliveryOptions;
      this.quantity = quantity;
      this.isProductInStock = isProductInStock;
      this.archived = archived;
      this.sellCount = sellCount;
      this.reviewCount = reviewCount;
      this.likesCount = likesCount;
      this.commentCount = commentCount;
      this.timestamp = timestamp;
    }
    get priceInCents():number {return this.price * 100;}
    
    //convert postModel to Map
    toMap() {
      return {
        postId: this.postId,
        ownerId: this.ownerId,
        username: this.username,
        description: this.description,
        productName: this.productName,
        brand: this.brand,
        variants: this.variants,
        price: this.price,
        location: this.location,
        rating: this.rating,
        images: this.images,
        selectedVariant: this.selectedVariant,
        delivery: this.deliveryOptions,
        quantity: this.quantity,
        isProductInStock: this.isProductInStock,
        archived: this.archived,
        sellCount: this.sellCount,
        reviewCount: this.reviewCount,
        likesCount: this.likesCount,
        commentCount: this.commentCount,
        timestamp: this.timestamp,
      };
    }

    toAlgolia() {
      return {
        "postId": this.postId,
        "ownerId": this.ownerId,
        'objectID': this.postId,
        "productName": this.productName,
        "brand": this.brand,
        'username': this.username,
        "price": this.price,
        "description": this.description,
        "location": this.location,
        "archived": this.archived,
        "sellCount": this.sellCount,
        "timestamp": this.timestamp,
      };
    }
}